Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs

```diff
@@ -48,19 +48,36 @@ internal static HttpRulePathPattern Parse(string pattern, MessageDescriptor requ
                     endOfName = match.Value.Length - 1;
                 }
                 string propertyPath = match.Value.Substring(1, endOfName - 1);
-                segments.Add(CreatePropertyAccessor(requestDescriptor, propertyPath));
+                segments.Add(CreatePatternSegment(requestDescriptor, propertyPath));
                 lastEnd = match.Index + match.Length;
             }
             if (lastEnd != pattern.Length)
             {
                 string literal = pattern.Substring(lastEnd);
-                segments.Add(message => literal);
+                segments.Add(HttpRulePathPatternSegment.CreateLiteral(literal));
             }
             return new HttpRulePathPattern(segments);
         }
 
         // TODO: do we need to perform URL encoding of anything?
-        internal string Format(IMessage message) => string.Join("", _segments.Select(segment => segment(message)));
+        internal string Format(IMessage message) => string.Join("", _segments.Select(segment => segment.Format(message)));
+
+        /// <summary>
+        /// Names of the fields of the top-level message that are bound by this pattern.
+        /// </summary>
+        internal List<string> TopLevelFieldNames => _segments
+            .Where(segment => segment.TopLevelFieldName != null)
+            .Select(s => s.TopLevelFieldName)
+            .ToList();
+
+        private static HttpRulePathPatternSegment CreatePatternSegment(MessageDescriptor descriptor, string propertyPath)
```

Comment:
Move into HttpRulePathPatternSegment, like the other factory methods? (The HttpRulePathPatternSegment constructor could then be private, I suspect.)

Reply:
👍